### ISSUE TITLE
Copy/replace/add left <-> right in compare views

### ISSUE THREAD

**Issue by Dominik-K** [microsoft/vscode#25887]
In the compare views, it would be nice to have 

* "Copy to left/right"
* "Replace left (with right)/right (with left)"
* "Add above/down left/right"

in *all* views, i.e. the side-by-side and the unified view (there with other text than "left/right").

Now, I need to select the text and copy-paste it. In the unified view, I even can't select the text from the compared file ("left" file in side-by-side view).

A nice UI would be the one as in [Meld](http://meldmerge.org/):
![image](https://cloud.githubusercontent.com/assets/5795847/25697422/2cdf426e-30bb-11e7-91ba-af67bdebe88d.png)

Related: #24390 (but only wanted for staging VCS diffs)
Blocked by: https://github.com/Microsoft/vscode/issues/4180 ?

**Comment by gulshan** [comment]
Similar proposal by me following git merge inline actions- #30519

**Comment by elgrafico** [comment]
I miss that feature as well. Good old [NetBeans ](https://ui.netbeans.org/docs/ui/DiffSpecification/DIFFSpecification.html#Figure2)has a a similar intuitive approach:

  ![diff_netbeans-8 2](https://user-images.githubusercontent.com/754317/34676314-b4100e72-f48c-11e7-8af6-bd0dcdfdea40.png)

Only the right side can be edited though, but there's a "swap" button right above the diff to mitigate that ;-)

**Comment by otienoanyango** [comment]
+1 This would be really useful to have. Any hope of this

**Comment by rkotenko** [comment]
This is probably my biggest problem with VS Code so far.  We are trying to drop Intellij, which lets you compare any two files and copy right and left with ease.

**Comment by bradkemper** [comment]
I was very surprised that this did not already exist in VS Code. BBEdit isn't perfect, but it does have that.

**Comment by seanchann** [comment]
+1

**Comment by MaerF0x0** [comment]
Super bonus points for allowing mapping to keys so i can create a vim-diff copy 
(eg:  `do`  takes the other (usually left) side's changes)

**Comment by sigmarising** [comment]
This is a really good feature which I am looking forward to.

In most coding time, I need to diff and merge only two files. I just want to diff and merge and there is no need for me to use git diff for only two files.

In present, I use the tool "diff & merge" from jet brains IDE. This is a really good sample plate to reference.

I strongly suggest this feature provided in the future version of VSCode.

![The diff & merge feature in JetBrains IDE](https://user-images.githubusercontent.com/25470568/43189752-ee03d6de-9029-11e8-8fdb-21e19a056492.gif)

**Comment by vbuyonje** [comment]
Many of us awaiting why not write it

**Comment by gongiskhan** [comment]
+1

**Comment by C0smin** [comment]
We need this so bad

**Comment by ufchta** [comment]
This is probably my biggest problem with VS Code so far. I was very surprised that this did not already exist.
Any chance that this will be added?

**Comment by andreasanta** [comment]
+1

**Comment by gztomas** [comment]
+1

**Comment by maxthemillion** [comment]
+1

**Comment by matt-wormley** [comment]
+1

**Comment by gulshan** [comment]
As my proposal #30519 was closed as a duplicate of this one, pasting the text here-

Inline actions in diff view for copy, replace, previous and next
=
This proposal requests inline actions in diff view, like the inline merge actions are provided for git merge markers within the editor-
![inline_diff](https://user-images.githubusercontent.com/751732/45917396-03a12900-be98-11e8-9683-9c6522c42134.png)

The code actions can be `Copy to left/right/other side`, `Replace left/right/other side`, `Previous` and `Next` (for going to previous/next diff). And as copying original is prohibited in inline diff view, there can be a `Copy original` action as well.

This proposal can be regarded as an implementation detail of the proposal #25887 .

**Comment by vvelkovski** [comment]
+1

**Comment by sanyamsingh** [comment]
+1

**Comment by moronito** [comment]
+1

**Comment by eamonsolvers** [comment]
+1

**Comment by limegreen-wayne** [comment]
+1

**Comment by DamienFlury** [comment]
+1

**Comment by Montago** [comment]
+1

**Comment by Montago** [comment]
Just a comment: I noticed that i could edit the "right side" of Compare ... that is, my local copy. 

This made it possible to take edits from the foreign file - but in the event i needed to sync the other way, i'd have to overwrite the foreign file with the local updated edit. 

Merging the two files and then have VSCODE overwrite both would be awesome.

**Comment by gabrial11392** [comment]
VS code need this.

**Comment by mainmachine** [comment]
I'm shocked no one has written an extension for this!

Meld is just so much better for this purpose.

**Comment by oca159** [comment]
+1

**Comment by SaschaKoth** [comment]
+1

**Comment by defernan** [comment]
+100000000000

**Comment by sloong** [comment]
+1

**Comment by sbrech** [comment]
+1

**Comment by jstewart-lgscout** [comment]
+1

**Comment by sashang** [comment]
I can diff 2 files but can't merge left or right? It's something I've been able to do in Vim for decades...

**Comment by a19i23** [comment]
This would be a great addition. I really like VSCode so far. If you add this functionality that would make me love VSCode!

**Comment by amarchan1** [comment]
Definitely needed

**Comment by dguera** [comment]
This needs to be added ASAP!

**Comment by xahon** [comment]
@Microsoft why you don't just make vscode open source? I think of it would be open sourced new useful features would appear more and more faster than waiting while you handle it

**Comment by Montago** [comment]
@xahon ... Guess why you are commenting on github ?

**Comment by simar88** [comment]
Any news about it? 👍

**Comment by zjjott** [comment]
+1
If this feature is supported,pls tell me

**Comment by Jan1torEarl** [comment]
+1
Was shocked to find this isn't already a feature... I thought I was doing something wrong at first!

**Comment by mightyx** [comment]
+1, this is missing

**Comment by sguillia** [comment]
How can I know when this gets implemented without receiving a notification every day?

**Comment by PNRxA** [comment]
Agreed, please thumb up the issue instead of leaving a +1 and needlessly notifying everyone who subscribes to this.

**Comment by nils-tekampe** [comment]
+1

**Comment by julien-conan** [comment]
+1

**Comment by medmehdim** [comment]
Ooooh thanks god ! just noticed i'm not the only one who's struggling.

**Comment by on-three** [comment]
Really liking visual studio code. Wonderful cross platform IDE, but it woefully needs this feature.

**Comment by jfsicilia** [comment]
Hope this feature won't delay.

**Comment by julien-conan** [comment]
Feature has been developed through VS Code Extension already, available at the Market Place : https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github

**Comment by bradkemper** [comment]
@julien-conan Are you sure that’s the right link? This issue is not about GitHub pull requests. Not even about Git merging.

**Comment by julien-conan** [comment]
@bradkemper ho ho! I was sure Something was wrong… hello guys :) ^^

**Comment by luxcium** [comment]
@xahon you don’t have any «fork» button ? I mean you either need to now you are sarcastic or trolling or maybe you don’t understand that to have the best thing in life everything must follow a complex process to make sure noting weak or unfit is included in this almost perfect IDE.

If you think the maintainer of this project is to strict on whom can collaborate remember that anything you can imagine about VScode can be provided as an extension created by anyone!!! If you ça write quality code and submit it it will probably be included in the community effort!!!

**Comment by dpinol** [comment]
+1

**Comment by nztraveler** [comment]
we really need this feature ASAP!
（救救孩子吧） o(╥﹏╥)o

**Comment by srameshr** [comment]
F

**Comment by B-iggy** [comment]
@alexandrudima It's the only feature missing for me to migrate from PHPStorm to Visual Code.
In your opinion: how difficult would it be to implement it? Can we somehow help?

**Comment by zhengminhui** [comment]
Can we add this feature in the 'Working Tree' view? So you can choose which part to change when you make local changes.

**Comment by nemchik** [comment]
I want to state a distinction that hopefully does not go unnoticed. VSCode already has git merge capabilities that meet my needs. What it lacks is the ability to diff/merge ad-hoc.

Right now I can open two random unrelated files and compare their differences (VSCode already has this) but i cannot merge differences left or right without manually editing the files.

So what I'm looking for is this interface:
![image](https://user-images.githubusercontent.com/725456/58108746-00d70f00-7bb2-11e9-9af8-90dfc0536333.png)
But add the ability to edit the file on the left (right now it's read-only) and buttons like meld or winmerge that let me chose which side to use and make both files match the chosen side.

**Comment by akurz42** [comment]
It's been 2 years since the OP - anybody still having hopes to see this anytime soon? It's the only thing that keeps me from switching from Atom.

**Comment by jraviotta** [comment]
+1

**Comment by alex-shamshurin** [comment]
+1

**Comment by yummip** [comment]
+1

**Comment by tonitone** [comment]
👍

**Comment by bryant-finney** [comment]
+1 

I am frequently examining the `diff` of two files, and it would be very nice if VSCode supported this functionality. It's annoying to require Meld or vimdiff for whenever I need to selectively copy changes

**Comment by wlarch** [comment]
+1 

As mentionned previously,  "diff & merge" from Jet Brains IDE works very well.

**Comment by magic-lantern** [comment]
Would also love to see this feature or an extension with this feature!

**Comment by wharehinga** [comment]
+ from me

**Comment by therevoman** [comment]
Please either let me edit the files in the diff view, or allow me to copy changes from one side to the other.

**Comment by xahon** [comment]
It's better to introduce new design than this useless feature

**Comment by cristi-neagu** [comment]
You can argue that there is a better way to do file compares, but as things are right now, this feature is most definitely not useless.

**Comment by BlakeTheAwesome** [comment]
I'm really surprised this isn't a feature of the compare view. It's probably the most common action to want to perform in that view.

**Comment by ipleten** [comment]
2 years...

**Comment by deyvidm** [comment]
I looked into writing an extension to help implement the feature. The biggest hurdle is that there's no way to add a handler for gutter decorators. Sure you can slide in an arrow icon when diffing two files, but there's no way to define behaviour on icon clicks. 

It seems very petty to withhold that functionality from the API, given the behaviour already exists in the implementation of breakpoints.

**Comment by evilrix** [comment]
I can't believe this is lacking as a feature. It's the same in Visual Studio! Without the ability to compare and merge changes, having a diff tool is almost pointless.

**Comment by avin-shum** [comment]
@deyvidm Just allowing hot-key to move changes is good enough.

**Comment by Easy-Tutor** [comment]
This is stoll not a feature in VS Code?

**Comment by Easy-Tutor** [comment]
This is still not a feature? :o

**Comment by moshfeu** [comment]
1. It will be nice to support it in the diff window opened by the `vscode.commands.executeCommand('vscode.diff')`

1. More advanced. Add an API to get diff details for example, is active line is not exists in the other window, if it partial different, what's the value in the other window, navigation between diffs etc.

**Comment by newtack** [comment]
This feature is definitely needed. Any update?

**Comment by jfscoertzen** [comment]
I would love this please. Then I don't have to use another tool.

**Comment by elieobeid7** [comment]
This should be among the most critical tools in the pipeline. There's even no plugin, if there was a vscode extension, it would be awesome.

**Comment by xhwang-chromium** [comment]
+1000 Please add this feature!

**Comment by mariusgaldigital** [comment]
+1 from me too. 
After the great update today with many new features this is highly anticipated 🤘

**Comment by bedge** [comment]
For the love of ... Argh, I though I was done with still needing use vim for this.

**Comment by Somnium7** [comment]
This is no longer blocked by #4180, as that issue is resolved. Will we have any progress on this one?

**Comment by SilverLinings89** [comment]
This issue keeps some members of our team on WebStorm :(
+1

**Comment by evilrix** [comment]
I can't believe this issue still isn't resolved. I keep getting VSCode updates with frivolous changes that I really don't care about (such as tinkering with the colours of the side bar and other features that are nice to have, but really have no effect on my workflow) and yet this really serious omission is still being left fixed.

If VSCode is to really be a contender as a best of breed editor I should be able to perform all my code editing workflow without having to leave it or use other editors. As it stands, I can edit code, I can commit code, I can even see what has changed, but there is no way to do anything with what I can see (in a useful way). I still have to have a seperate editor or diff tool as part of my workflow and this is just crazy!

I know VSCode is free. I know we're at liberty to use any editor we want, but the thing is VSCode really is just about the best code editor (especially for C++) around. The fact it has so many other features built right in makes it a joy to use. Intellisense makes my C++ coding experience about as good as it gets in Visual Studio. The problem is, this one omission spoils a workflow that is, for all intends and purposes, about as close to perfect as we're likely to get in a single editor.

Devs, please, please, please realise the importance of this omission. It's not just an inconvenience, it is a fundamental flaw. The fact there isn't even an extension to provide this feature makes it a serious fundamental flaw. That said, I really would prefer not to use an extension as these rarely have the polish nor integration that a baked in feature would. I assume the devs of VSCode actually use VSCode to develop VSCode, right? How have they not realised just how important this feature (or lack there of) is?

I'm at a point where I feel it may be necessary to dump VSCode and move on to something else. In the past I've tried Atom, but I've heard it's been greatly improved since the last time I tried it. I would be willing to lose the Intellisense (a nice to have) in favour of the ability to perform proper diff/merge tasks (a need to have for any developer). I truly hope the VSCode devs read this post and decide to finally take action on this as I really do not want to have to jump ship at this point.

Thank you.

**Comment by jrial** [comment]
People... We get it: you have really strong feelings about this issue, as do I. Please upvote the issue as described in the [contribution guidelines](https://github.com/microsoft/vscode/blob/master/CONTRIBUTING.md), but stop with the +1 comments/rants. They aren't used for prioritising and only serve to drown out the actual informative comments.

Your comment isn't going to land on one of the developers' desks and convince them to prioritise this. In fact, it's very likely they don't read the comments at all until after they already decided to start working on it. The only people you're pestering is your peers who already agree with you. That's 83 folks, at the time of writing.

Think before you comment. Thank you.

**Comment by evilrix** [comment]
Well, in all fairness the guidelines state: 

> make relevant comments and add your reaction

Explaining why this is so important is, IMHO, relevant! It's also nice to know that other people share your frustration. So, whilst I take your point on board, I don't think telling people to stop posting comments/rants is really the way to go. People like to be heard, even if it is only by their peers.

**Comment by moshfeu** [comment]
I created an extension to enable compare and merge changes. It still in alpha stage so there are not quite gaps and glitches but I would love to get feedbacks and improve it.

https://marketplace.visualstudio.com/items?itemName=moshfeu.diff-merge
https://github.com/moshfeu/vscode-diff-merge

**Comment by rcpao-enmotus-com** [comment]
At the very least, allow us to specify an external command to do the comparison: WinDiff, WinMerge, Meld, etc.

**Comment by mainmachine** [comment]
> At the very least, allow us to specify an external command to do the comparison: WinDiff, WinMerge, Meld, etc.

There are a couple of extensions to integrate Meld that work well, but don't really fulfill the goal of this feature request:

https://github.com/danielroedl/vsc-meld-diff
https://github.com/Aghabeiki/gitdiffer

**Comment by RedWingBB** [comment]
This would be an important feature in that select/copy/switch windows, paste,  can be difficult to manage when the source and destination is unfamiliar code as in a team situation

You can use WinMerge, which has this feature, but - WinMerge lacks Intellisense, and wont automatically check out the file being modified
This link is a little dated but you get the idea
http://blog.paulbouwer.com/2010/01/31/replace-diffmerge-tool-in-visual-studio-team-system-with-winmerge/

**Comment by jernster-clgx** [comment]
Adding my desperate request for this functionality to be built into VS Code.

**Comment by ejazrasool** [comment]
> People... We get it: you have really strong feelings about this issue, as do I. Please upvote the issue as described in the [contribution guidelines](https://github.com/microsoft/vscode/blob/master/CONTRIBUTING.md), but stop with the +1 comments/rants. They aren't used for prioritising and only serve to drown out the actual informative comments.
> 
> Your comment isn't going to land on one of the developers' desks and convince them to prioritise this. In fact, it's very likely they don't read the comments at all until after they already decided to start working on it. The only people you're pestering is your peers who already agree with you. That's 83 folks, at the time of writing.
> 
> Think before you comment. Thank you.

+1

**Comment by penx** [comment]
Related feature request - https://github.com/microsoft/vscode/issues/94365

**Comment by mrcw** [comment]
+1 :)

**Comment by zhangchenkai** [comment]
For 4 years, officials or 3rd extensions still not fully implement the function as  @sigmarising shown. Orz...

**Comment by joeyfezster** [comment]
It's been over 3 years since this was opened, any news?

**Comment by smarko82** [comment]
Any news on this?

**Comment by j12y** [comment]
I upvoted this issue because I'm spending a bit of time looking for a solution for this type of workflow.

Use Case: I have two files to compare that come from two separate git repositories.  Some of the differences are intentional, others need to be reconciled.  I mention because some of the diff/merge tools seem targeted for git history specifically.

Desired features:
- Selection of Files for Comparison
- Syntax Highlighting remains in diff view
- Diff Highlighting of lines added, removed, or changed
- Side-by-side comparison for viewing both files
- Jump to Diff (next/previous)
- Accept / Reject a difference 
- Edit file(s) in diff view
- Keybindings or shortcuts to make changes, with consideration for MacBook Pro or tenkeyless keyboards without function keys 
- No bugs or side-effects

Some alternatives I'm considering in case its useful for others

Name | Source | SyntaxHi | DiffHi | Side-by-Side | Jump to Next | Accept/Reject | Editable | Keybindings | Bugs
------|---------|----------|------|-----|----|------|-----|----|----
| vimdiff | External | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️
| Select for Compare | Built-in | ✔️  | ✔️ | ✔️ | ✔️ | ❌ | ✔️ | ❌ | ✔️
| Partial Diff | Extension | ❌ | ✔️ | ✔️ | ✔️ | ❌ | ❌ | ❌ | ✔️
| Diff & Merge | Extension | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ❌ 

For example I'm defaulting to `vimdiff file1 file2` has `[c` and `]c` to jump between changes and `dp` and `do` to put or obtain differences from one file or the other to reconcile differences.  Doing `:syntax off` can toggle syntax highlighting colors for easier readability of diff highlighting.  Looking for an alternative though if it exists or can improve my productivity while using VS Code.

**Comment by moshfeu** [comment]
@j12y , I saw you mentioned Diff & Merge has bugs. I'd love to hear more about them, can you please open (an) issue(s)?

**Comment by kevinhaas** [comment]
Give Webstorm a shot. It's an IDE for power users with features like this, and lots of expected functionality is default, not janky plugins. vscode is lacking in many areas and the performance is also very meh compared to Webstorm.

**Comment by bedge** [comment]
> I upvoted this issue because I'm spending a bit of time looking for a solution for this type of workflow and expected others would already have had a solution.

@j12y I've gone down the identical path. It's hard to believe that something 30 years old is so functionally superior to the current offerings.
You didn't mention any of the Jet Brains offerings. They have an acceptable diff integration:
![intellij-diff](https://user-images.githubusercontent.com/499317/93686015-505c7200-fa68-11ea-9f50-4eb632ae7e26.png)


A bit off track here but the other things that keeps me tied to vim is the multi-highlight. VCS's similarly named extension does a passable job at it, but not quite the same, even if you can force the same key bindings.

**Comment by javierguzman** [comment]
Another one here wishing this simple feature.

**Comment by SprayCanMan** [comment]
VSCode doo doo

**Comment by jonathanvila** [comment]
Would love to have this feature too :+1:

**Comment by dsideb** [comment]
same here. And it was implemented as a native feature, source control tools could benefit from it as well.

**Comment by 5ulo** [comment]
+1 for this functionality. It is a need.

**Comment by mauserzjeh** [comment]
+1 for this feature, really needed

**Comment by WissenIstNacht** [comment]
+1 for this.

**Comment by koolcardi** [comment]
+1

**Comment by moshfeu** [comment]
> I created an extension to enable compare and merge changes. It still in alpha stage so there are not quite gaps and glitches but I would love to get feedbacks and improve it.
> 
> https://marketplace.visualstudio.com/items?itemName=moshfeu.diff-merge
> https://github.com/moshfeu/vscode-diff-merge

For the new comers, you might find it useful.

**Comment by jonathanvila** [comment]
Hi Mosh

I installed your extension.... thank you for your work.

But, that extension has its own diff, afaik, I mean I would like to have
that approach for any diff vscode does : compare 2 files, compare with
previous version, compare with branch, differences on changes not staged,
compare 2 branches, etc......

I assume It would involve changing which is the tool used to show
differences....but I could be wrong.

**Comment by Florin-Popescu** [comment]
Since this issue seems to be discussing feature requests from other diff tools, perhaps other people might be interested in my suggestion for manual alignment in #113357?

**Comment by AnrDaemon** [comment]
meld has 3-way merge, which is WAY more useful in many cases.

**Comment by jrial** [comment]
> meld has 3-way merge, which is WAY more useful in many cases.

This issue isn't about that though. Check if there's already a feature request for adding 3-way merge, and add a thumbs up to the issue. That's how you actually vote on issues, not by commenting; see section "Feedback" in the project's README. If there's no issue yet, create one.

**Comment by nickbouton** [comment]
I'm with @Bitbangr on this - not everyone is on Win and this would be fine as first-class extension. I would love to have line-by-line left-right replacement similar to BeyondCompare on Windows.

The GitHub Desktop client is definitely handy (and I personally prefer using it for full side-by-side comparison to the current diff view in VS Code), but it doesn't offer anything near this in terms of merging/overwriting functionality, at least on OS X. In conflict resolution mode (e.g. mid-`git rebase`) it basically just hands off entirely to your IDE if you want anything other than brute-force entire file replacement.

**Comment by cristi-neagu** [comment]
@ewerybody It's because the compare functionality is built in. Why use a completely different compare tool if there is a compare tool already built in? And sure, an extension would work (as long as it adds functionality to the existing compare tool, rather than build a new compare tool), but if you're going to do that, why not simply add that functionality to VS Code directly? Is there any downside **at all** to having this functionality built in?

**Comment by ewerybody** [comment]
@8itbangr & @nickbouton no one was suggesting anyone should be on Windows :) I was giving an example.

Yeah. P4Merge is also good. 3-pane 👍 and nice curves to indicate origin 👍

I actually only saw the diff-view in VS Code when I actually had a conflict. And I actually DID see things like the "Accept Current/Incoming ..." [suggested here](https://github.com/microsoft/vscode/issues/30519) (also posted above).
Id did seem rigid tho.

Well, yeah @cristi-neagu there IS something built-in and it seems unfinished. The downside is bloat and technical debt. I'd move all that to an extension. That's what these are for. So far I'm quite happy with VS Codes policy there. Atom is IMHO a little too granular in that regard.

I just see aaaall these comments  ... "Many of us awaiting why not write it" ....
And I wanted to add a single voice saying:
> **I'm fine without it**!

I think thats ok. Its also very very Ok for you to look forward to an implementation 👍

**Comment by jonathanvila** [comment]
In my case I would love to have this when I compare clipboard with one
file, or one branch with current one, or one previous version of the local
file with current working version..... And several more....

And I think the best approach is to have this built-in vs code. For me is
only adding from source to destination, not in both ways, but in every
compare view already showed in vs code.

**Comment by mversluys** [comment]
vim has a similar feature which is super handy https://vimhelp.org/diff.txt.html#copy-diffs

This has me wondering if there's a way to add `do` and `dp` to https://marketplace.visualstudio.com/items?itemName=vscodevim.vim

**Comment by jcimoch** [comment]
What's the status of this issue, isn't going to be the part of 2021 roadmap?

**Comment by bupthebroker** [comment]
This is indeed a feature that is direly needed inside VS code diff editor.

Although i am surprised that people are suggesting the Meld approach. Give WinMerge a try and you'll see what true awesomeness is.

Meld has several problems, there are various keys for "push to", "pull from", "push from", "pull to" etc, and you need to have selected the right pane for some key combo to work. None of that nonsense is required. 

WinMerge has nailed this. Alt-Right to move a block from left to right and Alt-Left for the opposite.

**Comment by tipoldo** [comment]
> > I created an extension to enable compare and merge changes. It still in alpha stage so there are not quite gaps and glitches but I would love to get feedbacks and improve it.
> > https://marketplace.visualstudio.com/items?itemName=moshfeu.diff-merge
> > https://github.com/moshfeu/vscode-diff-merge
> 
> For the new comers, you might find it useful.

Thank you for this extension, it is very helpfull. But the changes can be done just in one direction (left to right). Why not in both?

**Comment by Wick3dJack** [comment]
Also a right click compare with for explorer in the context menu would be cool.

**Comment by alanwilter** [comment]
I'm quite upset that VS Code does not have this feature yet. Everything else I use or used had that (vim, Eclipse, BBedit). And though I appreciate the efforts for an extension, it's not practical yet. It must be a built-in feature. This is going now for 4 years!!! Hey VS Code devs, where are we on that!????

**Comment by sammy2142** [comment]
I would like to see this too :)

**Comment by jaroslawhartman** [comment]
Any news regarding this feature? Looking at this thread, it's really highly anticipated‼️

**Comment by sarahcpage** [comment]
Good grief please add this! It's making me regret leaving IntelliJ!

**Comment by AWhetter** [comment]
Please can we avoid adding comments like "I want this" and "when are we getting this". This is quite a high profile issue, meaning that a lot of people will be subscribed to email updates on the issue and they will be getting an email with every comment.
Instead, I encourage you to add a thumbs up to the issue (as described here https://code.visualstudio.com/blogs/2016/08/19/goodbyeuservoice#_github-reactions) because that's what the VSCode team will be using to influence their decision on what to work on next rather than on the number of comments.

**Comment by carwyn** [comment]
Related: #5770 and #37350

**Comment by saegl5** [comment]
+1 really hoping for this feature

**Comment by wuyiwei** [comment]
+1 really hoping for this feature

**Comment by dermatz** [comment]
Its 2021 .. wtf where is this basic IDE Feature ? .. Damn MSFT....

**Comment by AnrDaemon** [comment]
> Its 2021 .. wtf where is this basic IDE Feature ? .. Damn MSFT....

While I understand your sentiments, I disapprove the way you expressed it.

**Comment by zaunere** [comment]
Wow, this still isn't in place?  This is one of the most basic features of comparing two files - forget source control, git, etc.  As it is now, I compare two files, and then... what...?  There's nothing I can do after comparing.

**Comment by cristi-neagu** [comment]
It is weird that over 4 years since this request was made, it's still open. And now it's in the top 10 most requested features.

**Comment by nemchik** [comment]
To be fair, there were probably many other issues in the top 10 over the past 4 years they have knocked out.

I'm hopeful that this gets attention soon.

**Comment by gamalgogo** [comment]
AT least let us know when are you going to present this feature

**Comment by Campion** [comment]
+1 from me too

**Comment by APerricone** [comment]
without this the comparison is useless

**Comment by dermatz** [comment]
> > Its 2021 .. wtf where is this basic IDE Feature ? .. Damn MSFT....
> 
> While I understand your sentiments, I disapprove the way you expressed it.

I disapprove your useless comment. Compare Tools are essential in a modern IDE. Realize, that MSFT doesn't developed this basic feature is damn hard. Why they do a great job on 96% and mess up on the last steps? 

So you can be a fanboy - it doesn't care me - but this is the truth.

**Comment by APerricone** [comment]
trying to be propositive, an alternative can be configure the keys similar to this:
![image](https://user-images.githubusercontent.com/22457168/129694997-4e5db30b-36d9-443e-9e53-556ce3fa7e36.png)
Then use copy and paste to copy the difference between the two sides, beside in this way you can do it without a mouse.

**Comment by braddurbinzeth** [comment]
> trying to be propositive, an alternative can be configure the keys similar to this:
> ![image](https://user-images.githubusercontent.com/22457168/129694997-4e5db30b-36d9-443e-9e53-556ce3fa7e36.png)
> Then use copy and paste to copy the difference between the two sides, beside in this way you can do it without a mouse.

That still comes with an annoyance: you have to scroll to the correct place on the other side (look at the line number before switching, switch over, go to that line, and then paste).

**Comment by baylessj** [comment]
Chiming in to say that I'm also looking for this functionality. The compare view is nice for very simple diffs but not being able to move changes from one file to another means I'm going to look elsewhere for a diff tool.

**Comment by marcoskv** [comment]
Looks like a must-have feature for a developer editor. Hope it will be implemented soon...

**Comment by APerricone** [comment]
it is not a "must-have", but without this the possibility to compare 2 files has no sense

**Comment by tarekahf** [comment]
+1 I believe this is an important feature. In Eclipse, it is out of the box.

**Comment by Hades32** [comment]
If you only compare working tree vs head, you can use this as a workaround:
* focus right side
* F7 to go to next change
* CTRL-ALT-Z to revert selected range

Far from perfect, but better than manual copy&paste

**Comment by jtrumpio** [comment]
Typical...  Microsoft leaving everyone high and dry again...

**Comment by dermatz** [comment]
And if you say something, you get turned on straight away, because the choice of words is not okay. I only said that this was a basic feature and nothing more. But over 22 little pants-shitting fanboys always think they have to nod everything and vote against comments. What is that supposed to mean?

On the other hand, how difficult can it be to simply answer officially here to the feature request and whether it will be included in the roadmap or not. Open source is all well and good, but there are more IDEs out there than this. If you want the community to contribute, you should at least give a feature outlook.

But no, no time ... you have to do with voting down all critical contributions.

**Comment by jtrumpio** [comment]
Yeah, noticed already downvoted.  Pants-shitting fanboys is the best way to describe them, indeed.

**Comment by iXmerof** [comment]
Still discussing this since 2017 and no chance to get this implemented? 
+1 for this one, it's essential.

**Comment by jtrumpio** [comment]
Can it be like MySQL bug 20786 that has spanned decades and has even had birthday cakes baked for it?

Maybe this one will just have to settle for downvoting twatwaffles.

**Comment by tkossak** [comment]
Looks like another basic feature missing. Vscode looked awesome from the outside, but when you actually seriously start working with it, it starts to annoy you more and more with missing features.

**Comment by tonoslav** [comment]
> Looks like another basic feature missing. Vscode looked awesome from the outside, but when you actually seriously start working with it, it starts to annoy you more and more with missing features.

For me this is the only one really missing not solvable with extensions, or is it ?

**Comment by Rchatru** [comment]
I have to admit that when I saw that this issue was from 2017 I figured it would be resolved by now, but anyway, we have to be fair and see that since then they have fixed many bugs and added many new features (although I still see this one as basic).

In my opinion, I also think there has been a lack of communication from the development team, I would have appreciated: "we are not considering adding this feature now", or "we are studying it"..., so at least we would know what to expect.

Besides that, do you know of any alternative or workaround to achieve this?

P.S. As indicated in a previous comment, I understand that there are many people subscribed to this thread, so I have to apologize, but I wanted to share my feelings on this issue.

**Comment by brianjmurrell** [comment]
> If you only compare working tree vs head, you can use this as a workaround:
> 
> * focus right side
> * F7 to go to next change
> * CTRL-ALT-Z to revert selected range
> 
> Far from perfect, but better than manual copy&paste

This doesn't work for me.  CTRL-ALT-Z doesn't even appear to be bound to anything as near as I can tell.  What command-palette function is this supposed to be invoking?

**Comment by marcusschiesser** [comment]
> If you only compare working tree vs head, you can use this as a workaround:
> 
> * focus right side
> * F7 to go to next change
> * CTRL-ALT-Z to revert selected range
> 
> Far from perfect, but better than manual copy&paste

That's a life saver - it's CMD-OPTION-Z on Mac BTW

**Comment by brianjmurrell** [comment]
> If you only compare working tree vs head, you can use this as a workaround:
> 
> * focus right side
> * F7 to go to next change
> * CTRL-ALT-Z to revert selected range
> 
> Far from perfect, but better than manual copy&paste

So to be clear, this cannot be used to revert lines to a previous commit?  Kind of like "undo"ing through the commit history for the given line(s)?

**Comment by andreagalle** [comment]
Is [this](https://github.com/microsoft/vscode/issues/144519) going to relief our pain?

**Comment by nemchik** [comment]
> Is [this](https://github.com/microsoft/vscode/issues/144519) going to relief our pain?
> 
> #144519

I assume you're referencing https://github.com/microsoft/vscode/issues/37350 which is listed in the UX section of https://github.com/microsoft/vscode/issues/144519 meaning the idea behind this issue is at least getting attention enough to be considered (maybe even developed).

**Comment by WebSpectrumCode** [comment]
As it looks like this feature is not coming any time soon, is there a way to use external diff tool like Meld instead of VS Code built in one?

**Comment by RedWingBB** [comment]
Checkout WinMerge - an oldie but goodie, recently upped to 64bit 
Its a good merge tool with file and folder compare, copy left and right
However, since its not integrated into VS, you lose features like "Go to definition", etc

https://www.givetheiron.se/setting-up-winmerge-as-the-visual-studio-merge-and-diff-tool

**Comment by database64128** [comment]
[Meld](https://meldmerge.org/) is a cross-platform alternative to WinMerge.

**Comment by LZL0** [comment]
This would be indeed very very nice.

**Comment by Rchatru** [comment]
> > Is [this](https://github.com/microsoft/vscode/issues/144519) going to relief our pain?
> > #144519
> 
> I assume you're referencing #37350 which is listed in the UX section of #144519 meaning the idea behind this issue is at least getting attention enough to be considered (maybe even developed).

Here, #146091 we can follow the evolution of the issue #37350. They are working on it and I think it could be a good opportunity to unlock, at least collaterally, our much needed feature. Let's hope so.

**Comment by schnort** [comment]
Strangely, if you change your diff view to 'inline view', you can right click on the '-' or '+' and make context choices to revert the change.

(Assuming you're getting there from clicking on the file with changes in the source control window)

**Comment by weidenka** [comment]
+1 from my side. I also see this as a very basic feature for an IDE.

**Comment by fengyinqiao** [comment]
Ease of use is much worse than Jetbrains

**Comment by ansaryansu** [comment]
+1

**Comment by olegshulyakov** [comment]
One of last reasons why I still use free JetBrains products time to time

**Comment by hediet** [comment]
@mingpan can you share your UX ideas for this?

**Comment by RedWingBB** [comment]
Have a  look at winmerge
Next diff / forward/backward
Copy left/right
Copy left/right and advance
Copy all /left/right
Ability to manually edit to take a partial difference

**Comment by Mingpan** [comment]
Sure. Restating the UX idea from the above FR I previously opened:

I think one possible way to do this with minimal changes to the existing editor experience is to place arrows in the margin. These arrows can be conveniently used to revert a change in the compare (side-by-side diff) view. It only works in one direction (left -> right). But it should be able to save one click for reverting changes and also solve the problem that delete change can't be reverted (#98096). Below is a prototype I made to show how it looks like:

![arrow-revert](https://user-images.githubusercontent.com/15076498/172622361-c3f051e5-d1a8-4042-b629-770bc2c4ab04.gif)

Please let me know what you think.

**Comment by JamesHutchisonCarta** [comment]
>>  If you only compare working tree vs head, you can use this as a workaround:

>>  focus right side
F7 to go to next change
CTRL-ALT-Z to revert selected range
Far from perfect, but better than manual copy&paste

> That's a life saver - it's CMD-OPTION-Z on Mac BTW

This is not working for me. Next / previous diff doesn't seem to select anything. That button combo, or any near it, do not copy over changes from the compared branch.

**Comment by hediet** [comment]
> I think one possible way to do this with minimal changes to the existing editor experience is to place arrows in the margin. These arrows can be conveniently used to revert a change in the
> compare (side-by-side diff) view. It only works in one direction (left -> right). But it should
> be able to save one click for reverting changes and also solve the problem that delete change 
can't be reverted (#98096). Below is a prototype I made to show how it looks like:
> 
> ![arrow-revert](https://user-images.githubusercontent.com/15076498/> 172622361-c3f051e5-d1a8-4042-b629-770bc2c4ab04.gif)
> 
> Please let me know what you think.

I like this! @alexdima @misolori what do you think?

**Comment by alexdima** [comment]
@Mingpan Looks good! PR welcome.

**Comment by schnort** [comment]
Are breakpoints important or even relevant in the diff window?

**Comment by misolori** [comment]
This is why I said:

> only minor issue (and not sure if it's even a real issue) 

This is the only drawback and other extensions use the gutter besides breakpoints as well, so it's something to keep in mind but not blocking.

**Comment by bpasero** [comment]
> Are breakpoints important or even relevant in the diff window?

Yes, the right hand side of a diff editor should behave identical to a normal editor, i.e. you can edit, use editor widgets within, add breakpoints, etc.

**Comment by Sebi2020** [comment]
> Yes, the right hand side of a diff editor should behave identical to a normal editor
@bpasero why?! It makes no sense to set a breakpoint in a compare view. While editing and goto definitions within is useful and practical, breakpoints aren't.

**Comment by schnort** [comment]
While I understand that’s desirable, I’ll settle for a functional diff editor that lets me revert changes in the local copy back to the original.  Or easily handle conflicting merges.   

I don’t want the perfect preventing any movement towards the functional.

**Comment by gulshan** [comment]
I would request to look into my suggestion again. It is based on inline actions, thus leaving the gutter- https://github.com/microsoft/vscode/issues/25887#issuecomment-423741911

**Comment by bpasero** [comment]
@Sebi2020 

> It makes no sense to set a breakpoint in a compare view.

Treating the right hand side of a compare editor like a normal editor was a design decision from the beginning with the main intent to allow for edits. But I see no reason why we would artificially reduce functionality from the right hand side without reasons. 

When I review my changes I like the fact that I can test a change in the debugger right from the compare view by adding a breakpoint. See this example where I want to test a change to a test I did:

![Recording 2022-06-12 at 06 36 44](https://user-images.githubusercontent.com/900690/173214658-fa0530d8-3597-4153-89e7-df21a26515ac.gif)

Btw in the design from the OP I see no problem with still having breakpoints in the editor gutter:

<img width="378" alt="image" src="https://user-images.githubusercontent.com/900690/173214674-489d8cb8-1ff4-49e3-8386-1170284ee8ff.png">

**Comment by Nahor** [comment]
@bpasero, actually in that last picture, the cross is not a breakpoint, it's to [remove the change](https://netbeans.apache.org/kb/docs/ide/git.html#_comparing_file_revisions). So the gutter is probably not usable to add breakpoints or other non-diff functionality (that said, I never used Netbeans so I could be wrong)


> But I see no reason why we would artificially reduce functionality from the right hand side without reasons.

... uh, to avoid an overloaded UI? Isn't that what Misolori essentially said?

> I can test a change in the debugger right from the compare view by adding a breakpoint.

I guess to each is own. I appreciate the option to edit the code to fix minor issues (e.g. typos) and running the app/test to make sure I didn't introduce a stupid bug. But if I'm at a point where I need a breakpoint, personally I'd rather have a fullscreen window where I can see the whole code instead of being limited to a narrow column (in that regard, I'd love have a quick/easy way to switch to the fullscreen editor from the diff view, with the cursor already moved to the same location as in the diff, I think that would be more useful overall than being able to place/remove breakpoints)

**Comment by hediet** [comment]
Implemented in #151826.
If the breakpoint thing is a blocker, we can still revert.

**Comment by hediet** [comment]
I'm just realizing now that #151826 only partially adresses this.

Please create a new feature request if the feature-set not covered in that PR is important for you.

**Comment by APerricone** [comment]
I am touched, I saw arrows on compare! after only 5 years

**Comment by heartacker** [comment]
> I am touched, I saw arrows on compare! after only 5 years

you have right arrow after 5 years but left[失望]

**Comment by RedWingBB** [comment]
Sweet!
How about taking this valuable change to the compare in VS 2022?

**Comment by tdegeus** [comment]
As far as I understand I can only merge to the right, not to the left? Is this in the pipeline?

**Comment by KamasamaK** [comment]
> As far as I understand I can only merge to the right, not to the left? Is this in the pipeline?

See #153620

**Comment by tarekahf** [comment]
I'm very happy with this change. Now I can merge to the right and hopefully the left will come soon!

**Comment by liftconfig** [comment]
Anyone know if there is a keyboard shortcut for the new update from left to right functionality?
